GoogleContainerTools/skaffold
# fix: enable node platform affinity definition for `dev`, `debug` and `run` modes in mixed node clusters

This PR automatically enables node platform affinity rules for `dev`, `debug` and `run` modes when Skaffold detects a mixed node cluster. 

**Testing instructions**:
* Set a single platform cluster as active kubernetes context.
* In the `examples/getting-started` project, run: 
```
skaffold dev --default-repo=k8s-skaffold
```
* Check that the pod doesn't have any node affinity definition. The output of following command should be empty.
```
kubectl get pod getting-started  -o=jsonpath='{.spec.affinity}'
```
* Now, change the active kubernetes context to a mixed node cluster (googlers can use [this](https://pantheon.corp.google.com/kubernetes/clusters/details/us-central1-a/mixed-arch-cluster/details?project=cloud-code-ide-testing) cluster).
* Again run: 
```
skaffold dev --default-repo=k8s-skaffold
```
* Check that the pod has platform node affinity definition. 
```
kubectl get pod getting-started  -o=jsonpath='{.spec.affinity}'
```
The output of the above command should be like:
```
{"nodeAffinity":{"requiredDuringSchedulingIgnoredDuringExecution":{"nodeSelectorTerms":[{"matchExpressions":[{"key":"kubernetes.io/os","operator":"In","values":["linux"]},{"key":"kubernetes.io/arch","operator":"In","values":["arm64"]}]}]}}}
```
